docs/src/modules/components/DemoPropsForm.tsx

```diff
@@ -91,75 +75,117 @@ type DataType<PropName> = {
   max?: number;
 };
 
-interface ChartDemoPropsFormProps<PropName extends string> {
+type NumberDataType = {
+  knob: 'number' | 'slider';
+  /**
+   * The default value to be used by the components.
+   */
+  defaultValue?: number;
+} & DefaultTypes &
+  Pick<ConditionalTypes, 'step' | 'min' | 'max'>;
+
+type SelectDataType = {
+  knob: 'select' | 'color';
+  /**
+   * The default value to be used by the components.
+   */
+  defaultValue?: string;
+} & DefaultTypes &
+  Pick<ConditionalTypes, 'options'>;
+
+type RadioDataType = {
+  knob: 'radio';
+  /**
+   * The default value to be used by the components.
+   */
+  defaultValue?: string;
+} & DefaultTypes &
+  Pick<ConditionalTypes, 'options' | 'labels'>;
+
+type SwitchDataType = {
+  knob: 'switch';
+  /**
+   * The default value to be used by the components.
+   */
+  defaultValue?: boolean;
+} & DefaultTypes;
+
+type InputDataType = {
+  knob: 'input';
+  /**
+   * The default value to be used by the components.
+   */
+  defaultValue?: string;
+} & DefaultTypes;
+
+type PlacementDataType = {
+  knob: 'placement';
+  /**
+   * The default value to be used by the components.
+   */
+  defaultValue?: Placement;
+} & DefaultTypes;
+
+export type DataType =
+  | NumberDataType
+  | SelectDataType
+  | RadioDataType
+  | SwitchDataType
+  | InputDataType
+  | PlacementDataType;
+
+type OptionsOrString<Data extends Record<string, DataType>> = {
+  [K in keyof Data]: Data[K] extends { options: readonly string[] }
+    ? Data[K]['options'] extends readonly (infer T)[]
+      ? T
+      : string
+    : string;
+};
+
+export type PropsFromData<Data extends Record<string, DataType>> = {
+  [K in keyof Data]: Data[K]['knob'] extends 'number' | 'slider'
+    ? number
+    : Data[K]['knob'] extends 'select' | 'color'
+      ? OptionsOrString<Data>[K]
+      : Data[K]['knob'] extends 'radio'
+        ? OptionsOrString<Data>[K]
+        : Data[K]['knob'] extends 'switch'
+          ? boolean
+          : Data[K]['knob'] extends 'input'
+            ? string
+            : Data[K]['knob'] extends 'placement'
+              ? Placement
+              : never;
+};
```

Comment:
@bernardobelchior thanks for the help. I've added the following to your solution to allow for types to correctly guessed for all knobs and knobs with `options`, eg: `options: ['a', 'b']` => `props.options: 'a'|'b'`

Reply:
We should not use `defaultValue` because it is not required. Though maybe it should be? 

If a prop has no default then the result type would be `{a: any, b: any}`, while the `knob` is required, so it pays of to be more declarative in this case

Reply:
> We should not use defaultValue because it is not required. Though maybe it should be?

If `defaultValue` is undefined, what's the expected type? For the number knob, for example, should the value in the props object be `number` or `number | undefined`?